File: souce/148_Sort_List.cpp
/*
148. Sort List Medium

Overview:
    
    Sort a linked list in O(nlogn) time using constant space complexity.

Solution:
    
    Merge sort without recursion.
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (!head || !(head->next))
            return head;
        
        ListNode* cur = head;
        int length = 0;
        while (cur) {
            length++;
            cur = cur->next;
        }
        
        ListNode preHead(0);
        preHead.next = head;
        ListNode* tail, * first, * second;
        for (int step = 1; step < length; step <<= 1) {
            tail = &preHead;
            cur = preHead.next;
            while (cur) {
                first = split(cur, step);
                second = split(cur, step);
                tail = merge(tail, first, second);
            }
        }
        return preHead.next;
    }

private:
    ListNode* split(ListNode*& cur, int step) {
        ListNode* res = cur;
        for (int i = 1; i < step && cur; i++)
            cur = cur->next;
        if (cur) {
            ListNode* tmp = cur;
            cur = cur->next;
            tmp->next = NULL;
        }
        return res;
    }
    
    ListNode* merge(ListNode* tail, ListNode* first, ListNode* second) {
        while (first && second) {
            if (first->val < second->val) {
                tail->next = first;
                first = first->next;
            } else {
                tail->next = second;
                second = second->next;
            }
            tail = tail->next;
        }
        if (first) {
            tail->next = first;
        } else {
            tail->next = second;
        }
        while (tail->next)
            tail = tail->next;
        return tail;
    }
};

File: souce/148 Sort List.py
'''
Overview:

    Sort a linked list in O(n log n) time using constant space complexity.

Solution:

    Bottom-up merge sort
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        
        cur, length = head, 0
        while cur:
            length += 1
            cur = cur.next
        
        dummy = ListNode(0)
        dummy.next = head
        
        step = 1
        while step < length:
            cur, tail = dummy.next, dummy
            while cur:
                a = cur
                b = self.split(cur, step)
                cur = self.split(b, step)
                tail = self.merge(a, b, tail)
            step *= 2
        
        return dummy.next
    
    def split(self, head, n):
        i = 1
        while i < n and head:
            head = head.next
            i += 1
        if not head:
            return None
        res, head.next = head.next, None
        return res
    
    def merge(self, a, b, tail):
        while a and b:
            if a.val < b.val:
                tail.next, a = a, a.next
            else:
                tail.next, b = b, b.next
            tail = tail.next
        tail.next = a or b
        while tail.next:
            tail = tail.next
        return tail